pilopt/src/lib.rs

```diff
@@ -586,25 +717,28 @@ fn remove_constant_witness_columns<T: FieldElement>(pil_file: &mut Analyzed<T>)
     substitute_polynomial_references(pil_file, constant_polys);
 }
 
-/// Identifies intermediate columns that are constrained to a single value, replaces every
-/// reference to this column by the value and deletes the column.
-fn remove_constant_intermediate_columns<T: FieldElement>(pil_file: &mut Analyzed<T>) {
+/// Inlines `col i = e` into the references to `i` where `e` is an expression with no operations.
+/// The reasoning is that intermediate columns are useful to remember intermediate computation results, but in this case
+/// the intermediate results are already known.
+fn inline_trivial_intermediate_polynomials<T: FieldElement>(pil_file: &mut Analyzed<T>) {
     let intermediate_polys = pil_file
         .intermediate_polys_in_source_order()
         .filter_map(|(symbol, definitions)| {
             let mut symbols_and_definitions = symbol.array_elements().zip_eq(definitions);
             match symbol.is_array() {
                 true => None,
                 false => {
-                    let ((name, poly_id), definition) = symbols_and_definitions.next().unwrap();
-                    match definition {
-                        AlgebraicExpression::Number(value) => {
+                    let ((name, poly_id), value) = symbols_and_definitions.next().unwrap();
+                    match value {
+                        AlgebraicExpression::BinaryOperation(_) | AlgebraicExpression::UnaryOperation(_) => {
+                            None
+                        }
+                        _ =>{
```

Comment:
Could you list all the alternatives here? If we ever introduce a new operation (not likely, but still...) we won't find it here.

Reply:
Done.